### PR TITLE
add ptr::from_{ref,mut}

### PR DESCRIPTION
**Pull request by RalfJung**
We have methods to avoid almost all `as` casts around raw pointer handling, except for the initial cast from reference to raw pointer. These new methods close that gap.

(I also moved `null_mut` next to `null` to keep the file consistently organized.)

r? libs-api

Tracking issue: https://github.com/rust-lang/rust/issues/106116

### CODE REVIEW

**Comment by RalfJung** [comment]
@rust-lang/libs-api how do we make progress on this PR? It's just been sitting here for 3 weeks now.

File: library/core/src/ptr/mod.rs [hunk 1]
```diff
@@ -663,25 +684,26 @@ where
...
+/// This is equivalent to `r as *const T`, but is a bit safer since it will never silently change
+/// type or mutability, in particular if the code is refactored.
+#[inline(always)]
+#[must_use]
+#[unstable(feature = "ptr_from_ref", issue = "999999")] // FIXME
+pub fn from_ref<T: ?Sized>(r: &T) -> *const T {
+    r
+}
+
+/// Convert a mutble reference to a raw pointer.
```
Comment (dtolnay) [inline library/core/src/ptr/mod.rs:698]:
```suggestion
/// Convert a mutable reference to a raw pointer.
```

**Review by dtolnay — APPROVED** [review]
Seems good to me. Please file a tracking issue.

**Comment by dtolnay** [comment]
```console
curl: (6) Could not resolve host: ci-mirrors.rust-lang.org
```

@bors retry

**Pull request merged**